django/db/models/fields/tuple_lookups.py

```diff
@@ -0,0 +1,314 @@
+import itertools
+
+from django.core.exceptions import EmptyResultSet
+from django.db.models.expressions import Cols, Func, Value
+from django.db.models.lookups import (
+    Exact,
+    GreaterThan,
+    GreaterThanOrEqual,
+    In,
+    IsNull,
+    LessThan,
+    LessThanOrEqual,
+)
+
+
+class Tuple(Func):
+    function = ""
+
+
+class TupleLookupMixin:
+    def get_prep_lookup(self):
+        self.check_tuple_lookup()
+        return super().get_prep_lookup()
+
+    def check_tuple_lookup(self):
+        assert isinstance(self.lhs, Cols)
+        self.check_rhs_is_tuple_or_list()
+        self.check_rhs_length_equals_lhs_length()
+
+    def check_rhs_is_tuple_or_list(self):
+        if not isinstance(self.rhs, (tuple, list)):
+            raise ValueError(
+                f"'{self.lookup_name}' lookup of '{self.lhs.field.name}' field "
+                "must be a tuple or a list"
+            )
+
+    def check_rhs_length_equals_lhs_length(self):
+        if len(self.lhs) != len(self.rhs):
+            raise ValueError(
+                f"'{self.lookup_name}' lookup of '{self.lhs.field.name}' field "
+                f"must have {len(self.lhs)} elements"
+            )
+
+    def check_rhs_is_collection_of_tuples_or_lists(self):
+        if not all(isinstance(vals, (tuple, list)) for vals in self.rhs):
+            raise ValueError(
+                f"'{self.lookup_name}' lookup of '{self.lhs.field.name}' field "
+                f"must be a collection of tuples or lists"
+            )
+
+    def check_rhs_elements_length_equals_lhs_length(self):
+        if not all(len(self.lhs) == len(vals) for vals in self.rhs):
+            raise ValueError(
+                f"'{self.lookup_name}' lookup of '{self.lhs.field.name}' field "
+                f"must have {len(self.lhs)} elements each"
+            )
+
+    def tuple_comparison(self, compiler, connection, lookup_class):
+        # e.g.: (a, b, c) == (x, y, z) as SQL:
+        # WHERE (a, b, c) = (x, y, z)
+        vals = [
+            Value(val, output_field=col.output_field)
+            for col, val in zip(self.lhs, self.rhs)
+        ]
+
+        lookup = lookup_class(Tuple(self.lhs), Tuple(*vals))
+        return lookup.as_sql(compiler, connection)
```

Comment:
All of this logic should take into account that rhs might not be a direct value (e.g. it could be expressions returning a tuple). See how the lookups are implemented with `rhs_is_direct_value`.

Reply:
Yeah... I'll need to look into this tomorrow and test this extensively. I know it's handled with the `SubqueryConstraint` for related lookups. Maybe we should remove `SubqueryConstraint` altogether and move it here to the lookup classes?

Reply:
> Maybe we should remove SubqueryConstraint altogether and move it here to the lookup classes?

It could be worth exploring yes, `SubqueryConstraint` like `MultiColSource` always felt like _hacks_ that were added due to the lack of proper lookup supports.